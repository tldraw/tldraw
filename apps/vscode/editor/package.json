{
	"name": "@tldraw/vscode-editor",
	"description": "An an editor for the tldraw vscode extension.",
	"version": "0.0.0",
	"private": true,
	"author": {
		"name": "tldraw Inc.",
		"email": "hello@tldraw.com"
	},
	"homepage": "https://tldraw.dev",
	"license": "MIT",
	"repository": {
		"type": "git",
		"url": "https://github.com/tldraw/tldraw"
	},
	"bugs": {
		"url": "https://github.com/tldraw/tldraw/issues"
	},
	"keywords": [
		"tldraw",
		"drawing",
		"app",
		"development",
		"whiteboard",
		"canvas",
		"infinite"
	],
	"scripts": {
		"build": "yarn run -T tsx scripts/build.ts",
		"dev": "yarn run -T tsx scripts/dev.ts",
		"clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
		"lint": "yarn run -T tsx ../../../internal/scripts/lint.ts"
	},
	"devDependencies": {
		"@tldraw/assets": "workspace:*",
		"@types/fs-extra": "^11.0.4",
		"@types/node": "~20.11.30",
		"@types/react": "^18.3.18",
		"@types/react-dom": "^18.3.5",
		"@types/react-router-dom": "^5.3.3",
		"concurrently": "^9.1.2",
		"create-serve": "1.0.1",
		"dotenv": "^16.4.7",
		"esbuild": "^0.24.2",
		"fs-extra": "^11.3.0",
		"lazyrepo": "0.0.0-alpha.27",
		"react": "^18.3.1",
		"react-dom": "^18.3.1",
		"tldraw": "workspace:*",
		"tslib": "^2.8.1"
	},
	"gitHead": "a7dac0f83ad998e205c2aab58182cb4ba4e099a6"
}
