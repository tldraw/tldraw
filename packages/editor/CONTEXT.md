```markdown
# CONTEXT.md - @tldraw/editor Package

This file provides comprehensive context for understanding the `@tldraw/editor` package, the core infinite canvas editor for tldraw.

## Package Overview

The `@tldraw/editor` package is the foundational layer of tldraw - a minimal infinite canvas editor without any specific shapes, tools, or UI. It provides the core editing engine that the main `tldraw` package builds upon.

**Key Distinction:** This package provides the editor engine only. For a complete editor with shapes and UI, use `@tldraw/tldraw` instead.

**Version:** 3.15.1  
**Compatibility:** Requires Node.js ^20.0.0, React ^18.0.0  
**Bundle:** Ships with separate CSS file (`editor.css`) that must be imported

## Architecture Overview

### Core Components

**1. Editor Class (`src/lib/editor/Editor.ts`)**
- Central orchestrator for all editor functionality
- Manages the store, camera, selection, tools, and rendering
- Implements the main API surface for programmatic interaction
- Event-driven architecture using EventEmitter3
- Reactive state management using `@tldraw/state` signals

**2. TldrawEditor Component (`src/lib/TldrawEditor.tsx`)**
- Main React component that renders the editor
- Handles store creation, loading states, error boundaries
- Manages editor lifecycle and mounting
- Provides theming (light/dark mode) and licensing

**3. Store Integration**
- Uses `@tldraw/store` for reactive data persistence
- Supports local IndexedDB persistence via `persistenceKey`
- Can accept external stores or create internal ones
- Handles loading states and sync status

### State Management Architecture

**Reactive Signals System:**
- Uses `@tldraw/state` for reactive state management
- Atoms for mutable state, Computed for derived state
- Automatic dependency tracking and efficient updates
- All editor state is reactive and observable

**Store Structure:**
- Document data stored in `TLStore` (shapes, pages, assets, etc.)
- Editor state (camera, selection, tools) stored separately
- Derivations compute dependent values efficiently
- History management with undo/redo support

### Tools and State System

**StateNode Architecture (`src/lib/editor/tools/StateNode.ts`)**
- Hierarchical finite state machine for tools
- Each tool is a StateNode with potential child states
- Event-driven with handlers for pointer, keyboard, tick events
- Supports both "branch" nodes (with children) and "leaf" nodes

**Tool Types:**
- Root state manages overall editor state
- Tool states handle specific user interactions
- Child states for complex tools (e.g., drawing, resizing)
- Configurable tool state charts

### Shape System

**ShapeUtil Architecture (`src/lib/editor/shapes/ShapeUtil.ts`)**
- Abstract base class for defining shape behavior
- Each shape type needs a corresponding ShapeUtil
- Handles rendering, geometry, interactions, and serialization
- Extensible system for custom shapes

**Key Shape Methods:**
- `getGeometry()` - Shape's geometric representation
- `component()` - React component for rendering
- `indicator()` - Selection indicator rendering
- `onResize()`, `onRotate()` - Interaction handlers

### Binding System

**BindingUtil Architecture (`src/lib/editor/bindings/BindingUtil.ts`)**
- Abstract base class for defining relationships between shapes
- Manages connections like arrows to shapes, text to shapes, etc.
- Handles binding creation, updates, and cleanup
- Each binding type needs a corresponding BindingUtil

**Key Binding Concepts:**
- Bindings connect shapes through relationships
- `fromId` and `toId` reference connected shapes
- BindingUtils define visual indicators and interaction behavior
- Automatically updated when connected shapes change

### Manager System

The editor uses specialized managers for different concerns:

**Core Managers:**
- `ClickManager` - Multi-click detection and handling
- `EdgeScrollManager` - Auto-scroll at viewport edges during interactions
- `FocusManager` - Focus state and keyboard event handling
- `FontManager` - Font loading and management
- `HistoryManager` - Undo/redo functionality  
- `ScribbleManager` - Brush/scribble interactions
- `SnapManager` - Shape snapping during interactions
- `TextManager` - Text measurement and rendering
- `TickManager` - Animation frame management
- `UserPreferencesManager` - User settings persistence

### Component System

**Default Components (`src/lib/components/default-components/`)**
- Minimal implementations for all editor UI elements
- Canvas, cursors, handles, selection indicators, grid
- Error fallbacks and loading screens
- Fully customizable via `components` prop

**Key Components:**
- `DefaultCanvas` - Main drawing surface
- `DefaultCursor` - Mouse cursor rendering
- `DefaultHandles` - Shape resize/rotate handles
- `DefaultSelectionBackground/Foreground` - Selection UI
- `DefaultGrid` - Viewport grid overlay

**Indicators System (`src/lib/components/default-components/DefaultSelectionBackground.tsx`)**
- Visual feedback for shape selection and interaction states
- Includes selection boxes, rotation handles, and resize handles
- Shape-specific indicators defined in ShapeUtil.indicator()
- Binding indicators for relationship visualization

### Text Editing Integration

**Tiptap Integration:**
- Uses `@tiptap/core` and related packages for rich text editing
- Provides collaborative text editing capabilities
- Handles text formatting, selection, and cursor management
- Integrates with tldraw's event system and state management

**Text Manager (`src/lib/editor/managers/TextManager.ts`):**
- Handles text measurement and font metrics
- Manages text input states and focus
- Coordinates with Tiptap editor instances
- Provides text layout and wrapping calculations

### Geometry and Math

**Primitive System (`src/lib/primitives/`)**
- `Vec` - 2D vector math
- `Mat` - 2D transformation matrices  
- `Box` - Axis-aligned bounding boxes
- Geometry2d classes for shape collision/intersection
- Comprehensive math utilities for canvas operations

**Geometry Classes:**
- `Rectangle2d`, `Circle2d`, `Polygon2d`, etc.
- Hit testing and intersection calculations
- Point-in-shape and shape-shape collision detection

### Event System

**Event Flow:**
1. DOM events captured by editor container
2. Processed through pointer/keyboard managers
3. Dispatched to current tool's StateNode
4. Tool updates editor state accordingly
5. Reactive system triggers re-renders

**Event Types:**
- Pointer events (down, move, up) with target detection
- Keyboard events with modifier key handling
- Wheel events for zooming/panning
- Tick events for animations

### Export and Serialization

**Export Capabilities:**
- SVG export with full shape fidelity
- PNG/JPEG export via canvas rendering
- Snapshot serialization for persistence
- Asset handling (images, videos, fonts)

**Deep Links:**
- URL-based state synchronization
- Camera position and selected shapes in URL
- Configurable deep link behavior

### Licensing and Watermark

**License Management:**
- Handles tldraw licensing and watermark display
- `LicenseProvider` and `LicenseManager` components
- Watermark removal with valid business license
- License validation and enforcement

## Key Files and Directories

### Core Implementation
- `src/lib/editor/Editor.ts` - Main editor class
- `src/lib/TldrawEditor.tsx` - React component wrapper
- `src/lib/config/createTLStore.ts` - Store creation and configuration
- `src/lib/options.ts` - Editor configuration options

### Tools and State
- `src/lib/editor/tools/StateNode.ts` - State machine base class
- `src/lib/editor/tools/RootState.ts` - Root state implementation
- `src/lib/editor/tools/BaseBoxShapeTool/` - Base tool for box shapes

### Shape System
- `src/lib/editor/shapes/ShapeUtil.ts` - Shape utility base class
- `src/lib/editor/shapes/BaseBoxShapeUtil.tsx` - Base for rectangular shapes
- `src/lib/editor/shapes/group/GroupShapeUtil.tsx` - Group shape implementation

### Binding System
- `src/lib/editor/bindings/BindingUtil.ts` - Binding utility base class

### Managers
- `src/lib/editor/managers/` - All manager implementations
- `src/lib/editor/managers/ClickManager.ts` - Multi-click handling
- `src/lib/editor/managers/EdgeScrollManager.ts` - Auto-scroll functionality
- `src/lib/editor/managers/FocusManager.ts` - Focus state management
- `src/lib/editor/managers/FontManager.ts` - Font loading and management
- `src/lib/editor/managers/HistoryManager.ts` - Undo/redo functionality
- `src/lib/editor/managers/ScribbleManager.ts` - Brush interactions
- `src/lib/editor/managers/SnapManager.ts` - Shape snapping
- `src/lib/editor/managers/TextManager.ts` - Text measurement and rendering
- `src/lib/editor/managers/TickManager.ts` - Animation frame management
- `src/lib/editor/managers/UserPreferencesManager.ts` - User settings

### Components and Indicators
- `src/lib/components/default-components/` - Default UI components
- `src/lib/components/default-components/DefaultCanvas.tsx` - Main drawing surface
- `src/lib/components/default-components/DefaultCursor.tsx` - Mouse cursor rendering
- `src/lib/components/default-components/DefaultHandles.tsx` - Shape handles
- `src/lib/components/default-components/DefaultSelectionBackground.tsx` - Selection indicators
- `src/lib/components/default-components/DefaultGrid.tsx` - Viewport grid

### Testing Infrastructure
- `src/test/TestEditor.ts` - Editor testing utilities and mocks
- `src/test/` - Integration tests and test helpers

### Utilities
- `src/lib/utils/` - Editor-specific utilities
- `src/lib/primitives/` - Math and geometry utilities
- `src/lib/hooks/` - React hooks for editor integration

## Development Patterns

### Creating Custom Shapes

```typescript
class MyShapeUtil extends ShapeUtil<TLMyShape> {
  static override type = 'my-shape' as const
  
  getGeometry(shape: TLMyShape) {
    return new Rectangle2d({
      width: shape.props.w,
      height: shape.props.h,
    })
  }
  
  component(shape: TLMyShape) {
    return <div>My Shape Content</div>
  }
  
  indicator(shape: TLMyShape) {
    return <rect width={shape.props.w} height={shape.props.h} />
  }
}
```

### Creating Custom Tools

```typescript
export class MyTool extends StateNode {
  static override id = 'my-tool'
  
  onEnter() {
    this.editor.setCursor({ type: 'cross', rotation: 0 })
  }
  
  onPointerDown(info: TLPointerEventInfo) {
    this.editor.createShape({
      id: createShapeId(),
      type: 'my-shape',
      x: info.point.x,
      y: info.point.y,
    })
  }
}
```

### State Management

```typescript
// Access reactive state
const selectedShapes = editor.getSelectedShapes()
const bounds = editor.getSelectionPageBounds()

// Subscribe to changes
editor.store.listen((entry) => {
  console.log('Store changed:', entry)
})

// Use transactions for atomic updates
editor.batch(() => {
  editor.createShape(shape1)
  editor.createShape(shape2)
  editor.selectAll()
})
```

### Testing Patterns

```typescript
import { TestEditor } from './test/TestEditor'

describe('MyFeature', () => {
  let editor: TestEditor
  
  beforeEach(() => {
    editor = new TestEditor()
  })
  
  it('should create shapes', () => {
    editor.createShape({ type: 'geo', id: ids.box1 })
    expect(editor.getOnlySelectedShape()).toBe(editor.getShape(ids.box1))
  })
})
```

## Dependencies

**Runtime Dependencies:**
- `@tldraw/state` ^3.15.1 - Reactive state management
- `@tldraw/state-react` ^3.15.1 - React integration for state
- `@tldraw/store` ^3.15.1 - Document store
- `@tldraw/tlschema` ^3.15.1 - Type definitions and migrations
- `@tldraw/utils` ^3.15.1 - Shared utilities
- `@tldraw/validate` ^3.15.1 - Schema validation
- `@tiptap/core` ^2.6.6 - Rich text editing foundation
- `@tiptap/react` ^2.6.6 - React integration for Tiptap
- `@tiptap/pm` ^2.6.6 - ProseMirror integration
- `@use-gesture/react` ^10.3.1 - Touch/gesture handling
- `classnames` ^2.5.1 - CSS class utility
- `core-js` ^3.39.0 - JavaScript polyfills
- `eventemitter3` ^5.0.1 - Event system
- `idb` ^8.0.0 - IndexedDB wrapper
- `is-plain-object` ^5.0.0 - Object type checking

**Key Peer Dependencies:**
- `react` ^18.0.0 - Required React version
- `react-dom` ^18.0.0 - Required React DOM version

## CSS and Styling

**CSS Bundle:**
- Ships with `editor.css` containing all core styles
- Uses CSS custom properties for theming
- Separate from tldraw.css (which includes shape styles)
- Must be imported: `import '@tldraw/editor/editor.css'`

**Styling Approach:**
- CSS-in-JS for dynamic styles (selections, cursors)
- Static CSS for layout and base component styles
- Theme variables for light/dark mode switching
- Minimal external styling dependencies

## Performance Considerations

**Reactive System Optimization:**
- Reactive system minimizes unnecessary re-renders through precise dependency tracking
- Computed values are cached and only recalculated when dependencies change
- Store changes are batched to prevent cascading updates
- Component re-renders are minimized through React memo and signal integration

**Rendering Performance:**
- Geometry calculations are cached and memoized using shape geometry cache
- Large shape counts handled via viewport culling - only visible shapes are rendered
- Canvas rendering optimized for 60fps interactions with efficient paint cycles
- SVG export uses virtualization for large documents
- Font loading is managed asynchronously to prevent layout shifts during text rendering

**Memory Management:**
- Unused shape utilities are garbage collected
- Event listeners are properly cleaned up on component unmount
- Large assets are handled with lazy loading and disposal
- Store history is pruned to prevent unbounded memory growth

## Extensibility Points

**Highly Customizable:**
- Shape definitions via ShapeUtil
- Binding definitions via BindingUtil
- Tool behavior via StateNode
- UI components via components prop
- Event handling via editor instance
- Styling via CSS custom properties

**Less Customizable:**
- Core editor logic and data flow
- Store structure and reactivity system
- Basic event processing pipeline
- Text editing integration with Tiptap
```