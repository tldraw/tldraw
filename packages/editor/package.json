{
	"name": "@tldraw/editor",
	"description": "A tiny little drawing app (editor).",
	"version": "3.6.0",
	"author": {
		"name": "tldraw Inc.",
		"email": "hello@tldraw.com"
	},
	"homepage": "https://tldraw.dev",
	"license": "SEE LICENSE IN LICENSE.md",
	"repository": {
		"type": "git",
		"url": "https://github.com/tldraw/tldraw"
	},
	"bugs": {
		"url": "https://github.com/tldraw/tldraw/issues"
	},
	"keywords": [
		"tldraw",
		"drawing",
		"app",
		"development",
		"whiteboard",
		"canvas",
		"infinite"
	],
	"/* NOTE */": "These `main` and `types` fields are rewritten by the build script. They are not the actual values we publish",
	"main": "./src/index.ts",
	"types": "./.tsbuild/index.d.ts",
	"style": "./editor.css",
	"/* GOTCHA */": "files will include ./dist and index.d.ts by default, add any others you want to include in here",
	"files": [
		"editor.css"
	],
	"scripts": {
		"test-ci": "lazy inherit",
		"test": "yarn run -T jest",
		"benchmark": "yarn run -T tsx ./internal/scripts/benchmark.ts",
		"test-coverage": "lazy inherit",
		"build": "yarn run -T tsx ../../internal/scripts/build-package.ts",
		"build-api": "yarn run -T tsx ../../internal/scripts/build-api.ts",
		"prepack": "yarn run -T tsx ../../internal/scripts/prepack.ts",
		"postpack": "../../internal/scripts/postpack.sh",
		"pack-tarball": "yarn pack",
		"lint": "yarn run -T tsx ../../internal/scripts/lint.ts"
	},
	"dependencies": {
		"@tldraw/state": "workspace:*",
		"@tldraw/state-react": "workspace:*",
		"@tldraw/store": "workspace:*",
		"@tldraw/tlschema": "workspace:*",
		"@tldraw/utils": "workspace:*",
		"@tldraw/validate": "workspace:*",
		"@types/core-js": "^2.5.5",
		"@use-gesture/react": "^10.2.27",
		"classnames": "^2.3.2",
		"core-js": "^3.31.1",
		"eventemitter3": "^4.0.7",
		"idb": "^7.1.1",
		"is-plain-object": "^5.0.0",
		"lodash.isequal": "^4.5.0"
	},
	"peerDependencies": {
		"react": "^18.2.0",
		"react-dom": "^18.2.0"
	},
	"devDependencies": {
		"@peculiar/webcrypto": "^1.4.0",
		"@testing-library/jest-dom": "^5.16.5",
		"@testing-library/react": "^15.0.6",
		"@types/benchmark": "^2.1.2",
		"@types/wicg-file-system-access": "^2020.9.5",
		"benchmark": "^2.1.4",
		"fake-indexeddb": "^4.0.0",
		"jest-canvas-mock": "^2.5.2",
		"jest-environment-jsdom": "^29.4.3",
		"lazyrepo": "0.0.0-alpha.27",
		"resize-observer-polyfill": "^1.5.1"
	},
	"jest": {
		"preset": "../../internal/config/jest/node/jest-preset.js",
		"testEnvironment": "../../../packages/utils/patchedJestJsDom.js",
		"fakeTimers": {
			"enableGlobally": true
		},
		"testPathIgnorePatterns": [
			"^.+\\.*.css$"
		],
		"moduleNameMapper": {
			"^~(.*)": "<rootDir>/src/$1",
			"\\.(css|less|scss|sass)$": "identity-obj-proxy"
		},
		"setupFiles": [
			"raf/polyfill",
			"jest-canvas-mock",
			"<rootDir>/setupTests.js"
		],
		"setupFilesAfterEnv": [
			"../../internal/config/setupJest.ts"
		]
	}
}
