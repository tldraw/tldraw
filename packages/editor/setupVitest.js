// Vitest setup file for editor package

// IndexedDB polyfill for browser storage testing
require('fake-indexeddb/auto')

// ResizeObserver polyfill for DOM observation testing
global.ResizeObserver = require('resize-observer-polyfill')

// Crypto polyfill for cryptographic operations
global.crypto ??= new (require('@peculiar/webcrypto').Crypto)()

// FontFace API polyfill for font testing
global.FontFace = class FontFace {
	load() {
		return Promise.resolve()
	}
}

// Document fonts API mock
document.fonts = {
	add: () => {},
	delete: () => {},
	forEach: () => {},
	[Symbol.iterator]: () => [][Symbol.iterator](),
}

// Text encoding/decoding polyfills
global.TextEncoder = require('util').TextEncoder
global.TextDecoder = require('util').TextDecoder

// Window fetch mock for network requests
window.fetch = async (input, init) => {
	return {
		ok: true,
		json: async () => [],
	}
}

// Window.matchMedia polyfill for media queries
// Use a conditional approach to avoid conflicts with jsdom
window.matchMedia = (query) => ({
	matches: false,
	media: query,
	onchange: null,
	addListener: () => {},
	removeListener: () => {},
	addEventListener: () => {},
	removeEventListener: () => {},
	dispatchEvent: () => {},
})

// Enhanced DOM API mocking for CSSStyleDeclaration
// This ensures all HTML elements have the required style methods
if (typeof CSSStyleDeclaration !== 'undefined') {
	if (!CSSStyleDeclaration.prototype.getPropertyValue) {
		CSSStyleDeclaration.prototype.getPropertyValue = function (property) {
			return this[property] || ''
		}
	}
	if (!CSSStyleDeclaration.prototype.setProperty) {
		CSSStyleDeclaration.prototype.setProperty = function (property, value) {
			this[property] = value
		}
	}
	if (!CSSStyleDeclaration.prototype.removeProperty) {
		CSSStyleDeclaration.prototype.removeProperty = function (property) {
			delete this[property]
		}
	}
}

// Mock PointerEvent for jsdom
if (typeof window !== 'undefined' && !window.PointerEvent) {
	global.PointerEvent = class PointerEvent extends Event {
		pointerId = 1
		pointerType = 'mouse'
		clientX = 0
		clientY = 0

		constructor(type, options = {}) {
			super(type, options)
			this.pointerId = options.pointerId || 1
			this.pointerType = options.pointerType || 'mouse'
			this.clientX = options.clientX || 0
			this.clientY = options.clientY || 0
		}
	}
}

// Mock DragEvent for jsdom
if (typeof window !== 'undefined' && !window.DragEvent) {
	global.DragEvent = class DragEvent extends Event {
		dataTransfer = {
			getData: () => '',
			setData: () => {},
			files: [],
		}

		constructor(type, options = {}) {
			super(type, options)
			this.dataTransfer = options.dataTransfer || {
				getData: () => '',
				setData: () => {},
				files: [],
			}
		}
	}
}

// Mock TouchEvent for jsdom
if (typeof window !== 'undefined' && !window.TouchEvent) {
	global.TouchEvent = class TouchEvent extends Event {
		touches = []

		constructor(type, options = {}) {
			super(type, options)
			this.touches = options.touches || []
		}
	}
}
