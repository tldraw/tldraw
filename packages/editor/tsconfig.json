{
	"extends": "../../internal/config/tsconfig.base.json",
	"exclude": ["node_modules", "dist", "**/*.css", ".tsbuild*"],
	"compilerOptions": {
		"outDir": "./.tsbuild",
		"rootDir": "src",
		"noImplicitOverride": false,
		"noImplicitReturns": false,
		"types": ["node", "@types/jest"]
	},
	"references": [
		{ "path": "../state" },
		{ "path": "../state-react" },
		{ "path": "../store" },
		{ "path": "../tlschema" },
		{ "path": "../utils" },
		{ "path": "../validate" }
	]
}
