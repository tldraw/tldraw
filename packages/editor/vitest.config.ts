/// <reference types="vitest" />
import { mergeConfig } from 'vitest/config'
import baseConfig from '../../internal/config/vitest/node-preset'

// editor package needs jsdom environment, CSS mocking, and global fake timers
export default mergeConfig(baseConfig, {
	test: {
		// Use jsdom environment like the original Jest setup
		environment: 'jsdom',
		// Setup files to replace setupFiles functionality
		setupFiles: ['../../internal/config/vitest/setup.ts', 'vitest-canvas-mock', './setupVitest.js'],
		// Global fake timers like Jest configuration
		fakeTimers: {
			toFake: ['setTimeout', 'clearTimeout', 'setInterval', 'clearInterval', 'Date'],
		},
		// Allow importing actual modules for partial mocking
		globals: true,
		// CSS module mocking (equivalent to identity-obj-proxy)
		css: {
			modules: {
				classNameStrategy: 'non-scoped',
			},
		},
	},
	// CSS and asset handling
	css: {
		modules: {
			// This provides similar functionality to identity-obj-proxy
			// CSS class names will be returned as-is for testing
			generateScopedName: (name) => name,
		},
	},
})
