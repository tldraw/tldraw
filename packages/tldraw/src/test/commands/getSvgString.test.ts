import { DefaultDashStyle, createShapeId } from '@tldraw/editor'
import { TestEditor } from '../TestEditor'

let editor: TestEditor

const ids = {
	boxA: createShapeId('boxA'),
	boxB: createShapeId('boxB'),
	boxC: createShapeId('boxC'),
}

const parser = new DOMParser()
function parseSvg({ svg }: { svg: string } = { svg: '' }) {
	return parser.parseFromString(svg, 'image/svg+xml').firstElementChild as SVGSVGElement
}

jest.useRealTimers()

beforeEach(() => {
	editor = new TestEditor()
	editor.setStyleForNextShapes(DefaultDashStyle, 'solid')
	editor.setStyleForSelectedShapes(DefaultDashStyle, 'solid')
	editor.createShapes([
		{
			id: ids.boxA,
			type: 'geo',
			x: 0,
			y: 0,
			props: {
				w: 100,
				h: 100,
				text: 'Hello world',
			},
		},
		{
			id: ids.boxB,
			type: 'geo',
			x: 100,
			y: 100,
			props: {
				w: 50,
				h: 50,
			},
		},
		{
			id: ids.boxC,
			type: 'geo',
			x: 400,
			y: 400,
			props: {
				w: 100,
				h: 100,
				fill: 'pattern',
			},
		},
	])
	editor.selectAll()
})

it('gets an SVG', async () => {
	const svg = await editor.getSvgString(editor.getSelectedShapeIds())

	expect(svg!.width).toBe(564)
	expect(svg!.height).toBe(564)
	expect(svg!.svg).toMatch(/^<svg/)
})

it('Returns all shapes when no ids are provided', async () => {
	const svg = parseSvg(await editor.getSvgString([]))

	const elm = document.createElement('wrapper')
	elm.appendChild(svg)

	expect(elm).toMatchSnapshot('All shapes')
})

it('Gets the bounding box at the correct size', async () => {
	const svg = await editor.getSvgString(editor.getSelectedShapeIds())
	const parsed = parseSvg(svg!)
	const bbox = editor.getSelectionRotatedPageBounds()!
	const expanded = bbox.expandBy(editor.options.defaultSvgPadding) // adds 32px padding

	expect(parsed.getAttribute('width')).toMatch(expanded.width + '')
	expect(parsed.getAttribute('height')).toMatch(expanded.height + '')
	expect(svg!.width).toBe(expanded.width)
	expect(svg!.height).toBe(expanded.height)
})

it('Matches a snapshot', async () => {
	const svg = parseSvg(await editor.getSvgString(editor.getSelectedShapeIds()))

	const elm = document.createElement('wrapper')
	elm.appendChild(svg)

	expect(elm).toMatchSnapshot('Basic SVG')
})

it('Accepts a scale option', async () => {
	const svg1 = (await editor.getSvgString(editor.getSelectedShapeIds(), { scale: 1 }))!

	expect(svg1.width).toBe(564)

	const svg2 = (await editor.getSvgString(editor.getSelectedShapeIds(), { scale: 2 }))!

	expect(svg2.width).toBe(1128)
})

it('Accepts a background option', async () => {
	const svg1 = parseSvg(
		await editor.getSvgString(editor.getSelectedShapeIds(), { background: true })
	)
	expect(svg1.style.backgroundColor).not.toBe('transparent')

	const svg2 = parseSvg(
		await editor.getSvgString(editor.getSelectedShapeIds(), { background: false })
	)
	expect(svg2.style.backgroundColor).toBe('transparent')
})
