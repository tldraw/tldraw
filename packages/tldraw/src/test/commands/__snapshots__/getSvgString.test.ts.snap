// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches a snapshot: Basic SVG 1`] = `
<wrapper>
  <svg
    class="tl-container tl-theme__force-sRGB tl-theme__light"
    data-color-mode="light"
    direction="ltr"
    height="564"
    stroke-linecap="round"
    stroke-linejoin="round"
    style="background-color: rgb(249, 250, 251);"
    viewBox="-32 -32 564 564"
    width="564"
    xmlns="http://www.w3.org/2000/svg"
  >
    <defs>
      <mask
        id="_export_4_rb_"
      >
        <rect
          fill="white"
          height="8"
          width="8"
          x="0"
          y="0"
        />
        <g
          stroke="black"
          stroke-linecap="round"
        >
          <line
            x1="0.6666666666666666"
            x2="2"
            y1="2"
            y2="0.6666666666666666"
          />
          <line
            x1="3.333333333333333"
            x2="4.666666666666666"
            y1="4.666666666666666"
            y2="3.333333333333333"
          />
          <line
            x1="6"
            x2="7.333333333333333"
            y1="7.333333333333333"
            y2="6"
          />
        </g>
      </mask>
      <pattern
        height="8"
        id="_export_4_ra__hash_pattern_light_0"
        patternUnits="userSpaceOnUse"
        width="8"
      >
        <rect
          fill="#fcfffe"
          height="8"
          mask="url(#_export_4_rb_)"
          width="8"
          x="0"
          y="0"
        />
      </pattern>
    </defs>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 0, 0)"
    >
      <path
        d="M0, 0L100, 0,100, 428,0, 428Z"
        fill="none"
        stroke="#1d1d1d"
        stroke-width="3.5"
      />
      <text
        alignment-baseline="mathematical"
        dominant-baseline="mathematical"
        fill="rgb(249, 250, 251)"
        font-family="'tldraw_draw', sans-serif"
        font-size="22"
        font-style="normal"
        font-weight="normal"
        stroke="rgb(249, 250, 251)"
        stroke-width="2"
      >
        <tspan
          alignment-baseline="mathematical"
          unicode-bidi="plaintext"
          x="16"
          y="-17"
        >
          Hello world
        </tspan>
      </text>
      <text
        alignment-baseline="mathematical"
        dominant-baseline="mathematical"
        fill="#1d1d1d"
        font-family="'tldraw_draw', sans-serif"
        font-size="22"
        font-style="normal"
        font-weight="normal"
      >
        <tspan
          alignment-baseline="mathematical"
          unicode-bidi="plaintext"
          x="16"
          y="-17"
        >
          Hello world
        </tspan>
      </text>
    </g>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 100, 100)"
    >
      <path
        d="M0, 0L50, 0,50, 50,0, 50Z"
        fill="none"
        stroke="#1d1d1d"
        stroke-width="3.5"
      />
    </g>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 400, 400)"
    >
      <path
        d="M0, 0L100, 0,100, 100,0, 100Z"
        fill="#494949"
      />
      <path
        d="M0, 0L100, 0,100, 100,0, 100Z"
        fill="url(#_export_4_ra__hash_pattern_light_0)"
      />
      <path
        d="M0, 0L100, 0,100, 100,0, 100Z"
        fill="none"
        stroke="#1d1d1d"
        stroke-width="3.5"
      />
    </g>
  </svg>
</wrapper>
`;

exports[`Returns all shapes when no ids are provided: All shapes 1`] = `
<wrapper>
  <svg
    class="tl-container tl-theme__force-sRGB tl-theme__light"
    data-color-mode="light"
    direction="ltr"
    height="564"
    stroke-linecap="round"
    stroke-linejoin="round"
    style="background-color: rgb(249, 250, 251);"
    viewBox="-32 -32 564 564"
    width="564"
    xmlns="http://www.w3.org/2000/svg"
  >
    <defs>
      <mask
        id="_export_2_r5_"
      >
        <rect
          fill="white"
          height="8"
          width="8"
          x="0"
          y="0"
        />
        <g
          stroke="black"
          stroke-linecap="round"
        >
          <line
            x1="0.6666666666666666"
            x2="2"
            y1="2"
            y2="0.6666666666666666"
          />
          <line
            x1="3.333333333333333"
            x2="4.666666666666666"
            y1="4.666666666666666"
            y2="3.333333333333333"
          />
          <line
            x1="6"
            x2="7.333333333333333"
            y1="7.333333333333333"
            y2="6"
          />
        </g>
      </mask>
      <pattern
        height="8"
        id="_export_2_r4__hash_pattern_light_0"
        patternUnits="userSpaceOnUse"
        width="8"
      >
        <rect
          fill="#fcfffe"
          height="8"
          mask="url(#_export_2_r5_)"
          width="8"
          x="0"
          y="0"
        />
      </pattern>
    </defs>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 0, 0)"
    >
      <path
        d="M0, 0L100, 0,100, 428,0, 428Z"
        fill="none"
        stroke="#1d1d1d"
        stroke-width="3.5"
      />
      <text
        alignment-baseline="mathematical"
        dominant-baseline="mathematical"
        fill="rgb(249, 250, 251)"
        font-family="'tldraw_draw', sans-serif"
        font-size="22"
        font-style="normal"
        font-weight="normal"
        stroke="rgb(249, 250, 251)"
        stroke-width="2"
      >
        <tspan
          alignment-baseline="mathematical"
          unicode-bidi="plaintext"
          x="16"
          y="-17"
        >
          Hello world
        </tspan>
      </text>
      <text
        alignment-baseline="mathematical"
        dominant-baseline="mathematical"
        fill="#1d1d1d"
        font-family="'tldraw_draw', sans-serif"
        font-size="22"
        font-style="normal"
        font-weight="normal"
      >
        <tspan
          alignment-baseline="mathematical"
          unicode-bidi="plaintext"
          x="16"
          y="-17"
        >
          Hello world
        </tspan>
      </text>
    </g>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 100, 100)"
    >
      <path
        d="M0, 0L50, 0,50, 50,0, 50Z"
        fill="none"
        stroke="#1d1d1d"
        stroke-width="3.5"
      />
    </g>
    <g
      opacity="1"
      transform="matrix(1, 0, 0, 1, 400, 400)"
    >
      <path
        d="M0, 0L100, 0,100, 100,0, 100Z"
        fill="#494949"
      />
      <path
        d="M0, 0L100, 0,100, 100,0, 100Z"
        fill="url(#_export_2_r4__hash_pattern_light_0)"
      />
      <path
        d="M0, 0L100, 0,100, 100,0, 100Z"
        fill="none"
        stroke="#1d1d1d"
        stroke-width="3.5"
      />
    </g>
  </svg>
</wrapper>
`;
